api/include/opentelemetry/common/spin_lock_mutex.h

```diff
@@ -68,8 +87,22 @@ class SpinLockMutex
         {
           return;
         }
-        // TODO: Issue PAUSE/YIELD instruction to reduce contention.
-        // e.g. __builtin_ia32_pause() / YieldProcessor() / _mm_pause();
+// Issue a Pause/Yield instruction while spinning.
+#if defined(_MSC_VER)
+        YieldProcessor();
+#elif defined(__i386__) || defined(__x86_64__)
+#  if defined(__clang__)
+        _mm_pause();
+#  else
+        __builtin_ia32_pause();
+#  endif
+#elif defined(__arm__)
+        // This intrinsic should fail to be found if YIELD is not supported on the current
+        // processor.
+        __yield();
+#else
+        // TODO: Issue PAGE/YIELD on other architectures.
```

Comment:
Could we put an `#error` here for unimplemented architecture?

For the slow path below at line 113 which says the goal is ~1000ns, but 1ms is 10^6 ns, perhaps typo here? And for `std::this_thread::sleep_for`, I think it makes little sense to sleep for less than one 1 CPU quantum slice which is usually 10ms or more.

Reply:
> Could we put an `#error` here for unimplemented architecture?

It's not an error, though, just a CPU power efficiency optimisation.

> For the slow path below at line 113 which says the goal is ~1000ns, but 1ms is 10^6 ns, perhaps typo here? And for `std::this_thread::sleep_for`, I think it makes little sense to sleep for less than one 1 CPU quantum slice which is usually 10ms or more.

Yeah, that's a typo.  The idea is to bump order of mangitue, I think `this_thread::yield` is also a big higher in actual ns cost.   I can update the comments, as for the sleep_for quantile, we have a benchmark, so I can show you the performance difference between 1ms + 10ms.   

Here's the output:
 ````
 BM_SpinLockThrashing/1/process_time/real_time                 0.171 ms        0.155 ms          804
BM_SpinLockThrashing/2/process_time/real_time                 0.658 ms        0.234 ms          200
BM_SpinLockThrashing/4/process_time/real_time                  1.37 ms        0.744 ms           84
BM_SpinLockThrashing/8/process_time/real_time                  2.92 ms        0.852 ms           55
BM_SpinLockThrashing/12/process_time/real_time                 3.74 ms         1.46 ms           32
BM_TenMsSleepSpinLockThrashing/1/process_time/real_time       0.179 ms        0.177 ms          707
BM_TenMsSleepSpinLockThrashing/2/process_time/real_time        1.55 ms        0.943 ms          116
BM_TenMsSleepSpinLockThrashing/4/process_time/real_time        2.91 ms         1.95 ms           48
BM_TenMsSleepSpinLockThrashing/8/process_time/real_time        4.92 ms         1.49 ms           21
BM_TenMsSleepSpinLockThrashing/12/process_time/real_time       22.3 ms         2.84 ms           11
```

Effectively using an entire quantum of sleep is likely to delay a bit too long.  Note: This is for a high-contention, low-lock time.  So I think the current 1ms sleep is a pretty good balance of throughput + CPU consumption, but I'm happy to toy around with other values if you'd like.

Reply:
How to read the output table here?

Reply:
added the table header.